indexer-app/int-test/cmr/indexer/test/services/index_fields.clj

```diff
@@ -0,0 +1,152 @@
+(ns cmr.indexer.test.services.index-fields
```

Comment:
I am not sure why you refactored the original `cmr.indexer.test.services.index-versions` namespace into this and other namespaces. The purpose of that test is to verify that we are saving the index versions correctly in elasticsearch when saving and deleting documents. The code is not dependent on the type of the index, so we don't need to repeat it for types other than collection.

Reply:
I refactored because our testing namespaces, in general, are a mess -- mostly due to accretion and non-separation into different namespaces. Separation into different namespaces does cause a penalty, though, when you have fixtures that you would only like to set up once. Using the approach I did here gives us the best of both worlds: separation of distinct items (concepts, in this case) and a single instantiation of the embedded Elasticsearch.

As for averting future mess, this refactoring provides a clear "framework"/example so that adding more integration tests here in the future (which we should do more of) will be very easy and clear for developers. Testing integration tests with only Elasticsearch running is soooo much faster than having to run through the rinse-and-repeat process of our sys-int tests. At a conservative effort, my testing for this change likely was 5-10 times faster than similar testing iterations I've done on sys-int.

Reply:
It's nice you are using `in-ns` to separate the code into different namespaces. What I am saying is that for what is being tested, the type of the concept is irrelevant and we don't need to test it for every single concept type.

I have mixed feelings with unit tests. What indexer does is trivial and having unit tests is mostly busy work that doesn't justify the effort. That's why we don't have unit tests for the indexing code and only use search integration tests to verify that things are indexed correctly. I like integration tests because it lists the use cases more clearly and the tests are tied to use cases, not to implementation details (where unit tests are often tightly coupled with).

Reply:
Quick check: all these are in the same namespace, that's what allows them to use the same test fixture.

Another check: there is only one change I made the the unit tests: after the change to support all revisions, a sing unit test failed.

The tests that I separated out into `./int-test` are integration tests (as defined [here](https://en.wikipedia.org/wiki/Software_testing#Integration_testing): tests between a running (embedded) Elasticsearch instance and our indexer API.  All of the tests added verify that searching Elastic using our indexer API actually works.

In particular, this change adds integration tests that use the `{:all-revisions-index? true}` option, creating passing tests that would fail if the changes I hadn't implemented weren't in the code. They do so in such a way that it is not necessary to test all the systems together in order to verify this.

Side note: the stuff we do in `sys-int` is closer to what is defined [here](https://en.wikipedia.org/wiki/Software_testing#System_testing) as "system testing".

Reply:
Oh, I should also add:  in part 3 of this work, the search-app will be updated, and that's when I will add `sys-int` tests.

Reply:
It may be useful to review the work here:
 * https://bugs.earthdata.nasa.gov/browse/CMR-4657

Reply:
Liu and I just had a super-great chat, clarifying a bunch of stuff -- we're both happy now :-)